pkg/filesystem/local_directory_unix.go

```diff
@@ -106,6 +106,15 @@ func (d *localDirectory) Link(oldName path.Component, newDirectory Directory, ne
 	})
 }
 
+func (d *localDirectory) Clonefile(oldName path.Component, newDirectory Directory, newName path.Component) error {
+	defer runtime.KeepAlive(d)
```

Comment:
Don't know what this does, just copying the pattern - can you educate me?

Reply:
To make sure that file descriptors don't leak when code forgets to call Close(), we have a finalizer set up. This means the Go garbage collector will call Close for us()... at some point.

This specific piece of code unboxes `d` and passes the containing file descriptor to `newDirectory.Apply()`. What happens if the garbage collector is run:
- after extracting `d.fd`, and
- before finishing the call to `newDirectory.Apply()`?

Answer: Go may see `d` is no longer live, meaning it closes `d.fd`. By putting the `runtime.Keepalive()` call at the top (which is just a dummy function that Go's garbage collector treats like a black box), we can ensure the finalizer isn't run before this function completes.

Reply:
Interesting, thanks 👍